document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -16,6 +17,19 @@
    */
   boolean upsert(Key key, Document document) throws IOException;
 
+  /**
+   * Upsert (create a new doc or update if one already exists) the given document into the doc store
+   * if condition is evaluated to true. Provides optimistic lock support for concurrency update.
+   *
+   * @param key Unique key of the document in the collection.
+   * @param document Document to be upserted.
+   * @param condition Filter condition to be evaluated, on success update the document
+   * @param isUpsert Optional parameter indicates insert or update, default is true
+   * @return True if success. False otherwise.
+   */
+  boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
```

Comment:
wondering if it might be better to create separate APIs instead one for insert and upsert ?

Reply:
Let's see the use case but here, compared to the existing upsert API, I thought of enhancing it with the optional parameter `isUpsert` if a client wants to explicitly control `insert` and `update`.
```
True indicates if the document doesn't exist, it will insert a new document. And, if a document exists, it will update.
False indicates if the document exists, update it otherwise don't do anything.
```
The default value for isUpsert will be `true`. Let me know, would this suffice?

Reply:
this works as well

Reply:
Since we are creating new APIs, I think it would be more cleaner to have separate `create` and `update` APIs. Thoughts?

Reply:
Adding this kind of parameters is usually a bad practice as it reflects a bad abstraction for our methods. Let's please have a different method instead of this.

Check 

> Programmer B feels honor-bound to retain the existing abstraction, but since isn't exactly the same for every case, they alter the code to take a parameter, and then add logic to conditionally do the right thing based on the value of that parameter.

in https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction for a illustratory reading :).

Reply:
I feel the methods can be separated on the true/false state of `isUpsert` var, so we would a new `upsert` method accepting filter and other will be an `update` method

Reply:
My proposal would be exactly what rishahb described. I don't think we could call an upsert method when we are actually not doing the upsert. Update and upsert are different use cases and my first question would be, do we actually need a `update` function?

Reply:
> Since we are creating new APIs, I think it would be more cleaner to have separate create and update APIs. Thoughts?
I was in thoughts of not adding additional API w/o having a use case as it will have to maintain. 

Semantically, we already have an upsert method which if a document exists update, and if doesn't exist create one.
We were in a need of having one which provides optimistic locking for concurrency. So, initially, I was thinking of just extending it with the condition.
```
boolean upsert(Key key, Document document, Filter condition)
```
@Suraj, will this suffice? If not, I will create two separate APIs.

Create API: Create a document it doesn't exist
```
boolean create(Key key, Document document)
```
Update API: update the document if the condition met
```
boolean update(Key key, Document document, Filter condition)
```

Reply:
Discussed offline with @jcchavezs.
The API 
```
boolean upsert(Key key, Document document, Filter condition)
```
solves the use cases required for concurrency, however, it bit inconsistent. For the insert, it ignores the condition. But, the client can expect otherwise that if the condition is not met, it should not create the document. So, we will go ahead with the second proposal of having two separate APIs - create and update with the condition.

```
boolean create(Key key, Document document)
boolean update(Key key, Document document, Filter condition)
```

Reply:
Yes this is much easier to follow

Reply:
Done. Added two new APIs.